README.md
# `echomap`

[![crates.io](https://img.shields.io/crates/v/echomap.svg)](https://crates.io/crates/echomap)
[![Build status](https://github.com/pjsier/echomap/workflows/CI/badge.svg)](https://github.com/pjsier/echomap/actions?query=workflow%3ACI)

Preview map files in the terminal

![Terminal recording gif](https://raw.githubusercontent.com/pjsier/echomap/main/img/recording.gif?raw=true)

## Installation

If you have `cargo` installed, you can run `cargo install echomap` and then run it from `$HOME/.cargo/bin`. More details on this are available in [`cargo-install` documentation](https://doc.rust-lang.org/cargo/commands/cargo-install.html).

There are also [binaries available](https://github.com/pjsier/echomap/releases) for MacOS, Windows and Linux.

## Usage

```
USAGE:
    echomap [FLAGS] [OPTIONS] <INPUT>

FLAGS:
    -a, --area       Print polygon area instead of boundaries
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -c, --columns <COLUMNS>        Sets the number of columns (in characters) of the printed output. Defaults to
                                   terminal height minus 1.
    -f, --format <FORMAT>          Input file format (tries to infer from file extension by default) [possible values:
                                   geojson, topojson, csv, shp, wkt, polyline, kml]
        --lat <LAT>                Name of latitude column (if format is 'csv')
        --lon <LON>                Name of longitude column (if format is 'csv')
        --precision <precision>    Precision value for polyline parsing
    -r, --rows <ROWS>              Sets the number of rows (in characters) of the printed output. Defaults to terminal
                                   width.
    -s, --simplify <simplify>      Proportion of removable points to remove (0-1 or 0%-100%) [default: 0.01]

ARGS:
    <INPUT>    File to parse or '-' to read stdin
```

## Contributing

See [CONTRIBUTING.md](./CONTRIBUTING.md)


CHANGELOG.md
# Changelog

## Unreleased

## [v0.7.3](https://github.com/pjsier/echomap/releases/tag/v0.7.3)

- Multiple dependency updates to fix failing installs

## [v0.7.2](https://github.com/pjsier/echomap/releases/tag/v0.7.2)

- Update to use Rust 2021 edition
- Multiple dependency updates

## [v0.7.1](https://github.com/pjsier/echomap/releases/tag/v0.7.1)

- Use `FeatureIterator` from [`geojson` v0.22.3](https://github.com/georust/geojson/releases/tag/0.22.3) to reduce memory consumption and runtime significantly for extremely large GeoJSON files

## [v0.7.0](https://github.com/pjsier/echomap/releases/tag/v0.7.0)

- Multiple dependency updates, including refactoring for `clap`

## [v0.6.2](https://github.com/pjsier/echomap/releases/tag/v0.6.2)

- Add test for shapefiles ([#21](https://github.com/pjsier/echomap/pull/21)) from [@niwa-s](https://github.com/niwa-s)
- Updated dependencies

## [v0.6.1](https://github.com/pjsier/echomap/releases/tag/v0.6.1)

- Updated dependencies
- Cleaned up newer clippy lints

## [v0.6.0](https://github.com/pjsier/echomap/releases/tag/v0.6.0)

- Added KML to supported file formats
- Updated `geo-types` to 0.7


src/map_grid.rs
use std::char;
use std::io::{self, Write};

use anyhow::{Context, Result};
use geo::algorithm::bounding_rect::BoundingRect;
use geo::algorithm::contains::Contains;
use geo::algorithm::intersects::Intersects;
use geo::algorithm::simplify_vw::SimplifyVw;
use geo::{GeoFloat, Geometry, Line, Point, Polygon, Rect};
use num_traits::FromPrimitive;
use rstar::{self, RTree, RTreeNum, RTreeObject, AABB};

const CELL_ROWS: i32 = 4;
const CELL_COLS: i32 = 2;

#[derive(Debug, Clone, PartialEq)]
pub enum GridGeom<T>
where
    T: GeoFloat + RTreeNum + FromPrimitive,
{
    Point(Point<T>),
    Line(Line<T>),
    Polygon(Polygon<T>),
}

impl<T> GridGeom<T>
where
    T: GeoFloat + RTreeNum + FromPrimitive,
{
    /// Simplify geometries into component pieces for GridGeom
    pub fn vec_from_geom(geom: Geometry<T>, simplification: T, is_area: bool) -> Vec<GridGeom<T>> {
        match geom {
            Geometry::Point(s) => vec![GridGeom::Point(s); 1],
            Geometry::MultiPoint(s) => s.into_iter().map(GridGeom::Point).collect(),
            Geometry::Line(s) => vec![GridGeom::Line(s); 1],
            Geometry::LineString(s) => s
                .simplify_vw(&simplification)
                .lines()
                .map(GridGeom::Line)
                .collect(),
            Geometry::MultiLineString(s) => s
                .into_iter()
                .map(|ls| ls.simplify_vw(&simplification))
                .flat_map(|ls| ls.lines().collect::<Vec<_>>())
                .map(GridGeom::Line)
                .collect(),
            Geometry::Polygon(s) => {
                if is_area {
                    vec![GridGeom::Polygon(s.simplify_vw(&simplification)); 1]
                } else {
                    s.simplify_vw(&simplification)
                        .exterior()
                        .lines()
                        .map(GridGeom::Line)
                        .collect()
                }
            }
            Geometry::MultiPolygon(s) => {
                if is_area {
                    s.simplify_vw(&simplification)
                        .into_iter()
                        .map(GridGeom::Polygon)
                        .collect()
                } else {
                    s.simplify_vw(&simplification)
                        .into_iter()
                        .flat_map(|p| p.exterior().lines().collect::<Vec<_>>())
                        .map(GridGeom::Line)
                        .collect()
                }
            }
            Geometry::Triangle(s) => vec![GridGeom::Polygon(s.to_polygon()); 1],
            Geometry::Rect(s) => vec![GridGeom::Polygon(s.to_polygon()); 1],
            Geometry::GeometryCollection(s) => s
                .into_iter()
                .flat_map(|g| GridGeom::<T>::vec_from_geom(g, simplification, is_area))
                .collect(),
        }
    }
}

impl<T> RTreeObject for GridGeom<T>
where
    T: GeoFloat + RTreeNum + FromPrimitive,
{
    type Envelope = AABB<[T; 2]>;

    fn envelope(&self) -> Self::Envelope {
        match self {
            GridGeom::Point(pt) => AABB::from_point([pt.x(), pt.y()]),
            GridGeom::Line(line) => {
                let bb = line.bounding_rect();
                AABB::from_corners([bb.min().x, bb.min().y], [bb.max().x, bb.max().y])
            }
            GridGeom::Polygon(poly) => {
                let bb = poly.bounding_rect().unwrap();
                AABB::from_corners([bb.min().x, bb.min().y], [bb.max().x, bb.max().y])
            }
        }
    }
}

/// Convert row/col coordinates to associated Braille hex value
pub fn braille_cell_value(row: i32, col: i32) -> u32 {
    match (row, col) {
        (0, 0) => 0x01,
        (0, 1) => 0x08,
        (1, 0) => 0x02,
        (1, 1) => 0x10,
        (2, 0) => 0x04,
        (2, 1) => 0x20,
        (3, 0) => 0x40,
        (3, 1) => 0x80,
        (_, _) => 0x00,
    }
}

/// Add the Braille offset base to the calculated cell value to generate a char
pub fn braille_char(suffix: u32) -> char {
    let braille_offset = 0x2800;
    char::from_u32(braille_offset + suffix).unwrap()
}

pub struct MapGrid<T>
where
    T: GeoFloat + RTreeNum + FromPrimitive,
{
    rows: i32,
    cols: i32,
    bbox: Rect<T>,
    cell_size: [f64; 2],
    inner_cell_size: [f64; 2],
    rtree: RTree<GridGeom<T>>,
}

impl<T> MapGrid<T>
where
    T: GeoFloat + RTreeNum + FromPrimitive,
{
    pub fn new(width: f64, height: f64, rtree: RTree<GridGeom<T>>) -> MapGrid<T> {
        let envelope = rtree.root().envelope();
        let bbox = Rect::new(envelope.lower(), envelope.upper());
        let box_width = bbox.width().to_f64().unwrap();
        let box_height = bbox.height().to_f64().unwrap();

        let box_aspect_ratio = box_width / box_height;
        let term_aspect_ratio = width / height;

        // Clamp dimensions to aspect ratio of the geometry bbox
        let (cols_f, rows_f) = match (
            term_aspect_ratio > 1.0,
            box_aspect_ratio > 2.0,
            term_aspect_ratio > (box_aspect_ratio * 2.0),
        ) {
            // Multiply or divide by 2.0 to account for columns being more narrow than rows
            (true, true, true) | (true, false, _) => (height * box_aspect_ratio * 2.0, height),
            (true, true, _) | (false, _, _) => (width, (width / box_aspect_ratio) / 2.0),
        };

        let cols = f64::ceil(width) as i32;
        let rows = f64::ceil(height) as i32;

        // Get dimensions of individual cells
        let cell_width = box_width / cols_f;
        let cell_height = box_height / rows_f;

        MapGrid {
            bbox,
            rows,
            cols,
            cell_size: [cell_width, cell_height],
            inner_cell_size: [
                cell_width / f64::from(CELL_COLS),
                cell_height / f64::from(CELL_ROWS),
            ],
            rtree,
        }
    }

    /// Iterate through cells, printing one line at a time
    pub fn print(&self) -> Result<()> {
        let stdout = io::stdout();
        let mut handle = io::BufWriter::new(stdout.lock());

        for r in 0..self.rows {
            let mut row_str = "".to_string();
            for c in 0..self.cols {
                let cell_value = self.query_cell_value(r, c);
                row_str.push_str(&braille_char(cell_value).to_string());
            }
            writeln!(handle, "{}", row_str).context("Error printing line")?;
        }
        Ok(())
    }

    // Get the minimum and maximum points of a cell
    fn min_max_points(
        &self,
        row: i32,
        col: i32,
        start_width: f64,
        start_height: f64,
        cell_size: [f64; 2],
    ) -> (Point<T>, Point<T>) {
        let min_x = start_width + (cell_size[0] * f64::from(col));
        let max_y = start_height - (cell_size[1] * f64::from(row));
        let max_x = T::from_f64(min_x + cell_size[0]).unwrap();
        let min_y = T::from_f64(max_y - cell_size[1]).unwrap();

        let min_pt = Point::new(T::from_f64(min_x).unwrap(), min_y);
        let max_pt = Point::new(max_x, T::from_f64(max_y).unwrap());
        (min_pt, max_pt)
    }

    // Determine whether any geometries within a cell intersect
    fn cell_intersects(&self, min_pt: Point<T>, max_pt: Point<T>) -> bool {
        let envelope = AABB::from_corners([min_pt.x(), min_pt.y()], [max_pt.x(), max_pt.y()]);
        let poly_bounds = Polygon::from(Rect::new(min_pt, max_pt));

        self.rtree
            .locate_in_envelope_intersecting(&envelope)
            .skip_while(|g| match g {
                GridGeom::Point(pt) => !poly_bounds.contains(pt),
                GridGeom::Line(line) => !poly_bounds.intersects(line),
                GridGeom::Polygon(poly) => !poly_bounds.intersects(poly),
            })
            .take(1)
            .next()
            .is_some()
    }

    /// For a given Braille 2x4 cell, query which cells have geometries in them
    fn query_cell_value(&self, row: i32, col: i32) -> u32 {
        let mut cell_value = 0x00;

        let bbox_min_x = self.bbox.min().x.to_f64().unwrap();
        let bbox_max_y = self.bbox.max().y.to_f64().unwrap();

        // Return early if there are no geometries in the outer cell
        let (outer_min_pt, outer_max_pt) =
            self.min_max_points(row, col, bbox_min_x, bbox_max_y, self.cell_size);
        if !self.cell_intersects(outer_min_pt, outer_max_pt) {
            return cell_value;
        }

        // Get the start offset dimensions based on the outer row and column
        let start_width = (self.cell_size[0] * f64::from(col)) + bbox_min_x;
        let start_height = bbox_max_y - (self.cell_size[1] * f64::from(row));

        for r in 0..CELL_ROWS {
            for c in 0..CELL_COLS {
                let (min_pt, max_pt) =
                    self.min_max_points(r, c, start_width, start_height, self.inner_cell_size);
                // Add the associated cell value if intersecting lines are found
                if self.cell_intersects(min_pt, max_pt) {
                    cell_value += braille_cell_value(r, c);
                }
            }
        }

        cell_value
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use float_cmp::approx_eq;
    use geo_types::LineString;
    use num_traits::cast::ToPrimitive;

    #[test]
    fn test_vec_from_geom() {
        let poly = Polygon::new(LineString::from(vec![(0., 0.), (1., 1.), (1., 0.)]), vec![]);
        assert_eq!(
            GridGeom::<f64>::vec_from_geom(Geometry::Polygon(poly.clone()), 0.01, false),
            vec![
                GridGeom::Line(Line::<f64>::new((0., 0.), (1., 1.))),
                GridGeom::Line(Line::<f64>::new((1., 1.), (1., 0.))),
                GridGeom::Line(Line::<f64>::new((1., 0.), (0., 0.))),
            ]
        );
        assert_eq!(
            GridGeom::<f64>::vec_from_geom(Geometry::Polygon(poly.clone()), 0.01, true),
            vec![GridGeom::Polygon(poly)]
        );
    }

    #[test]
    fn new_clamps_aspect_ratio() {
        let line = GridGeom::Line(Line::new([0., 0.], [5., 1.]));
        let rtree = RTree::bulk_load(vec![line]);
        let grid = MapGrid::new(4., 4., rtree);
        assert_eq!((grid.cols, grid.rows), (4, 4));
        assert!(approx_eq!(f64, grid.cell_size[0], 1.25));
        assert!(approx_eq!(f64, grid.cell_size[1], 2.5));
    }

    #[test]
    fn query_cell_value_returns_value() {
        let rtree = RTree::bulk_load(vec![
            GridGeom::Line(Line::new([0., 0.], [4., 0.])),
            GridGeom::Point(Point::new(0., 1.)),
        ]);
        let grid = MapGrid::new(4., 4., rtree);
        assert_eq!(grid.query_cell_value(0, 0), 0x36);
    }

    #[test]
    fn min_max_points() {
        let rtree = RTree::bulk_load(vec![
            GridGeom::Line(Line::new([0., 0.], [4., 0.])),
            GridGeom::Point(Point::new(0., 1.)),
        ]);
        let grid = MapGrid::new(4., 4., rtree);
        let (row, col) = (0, 0);
        let start_width =
            (grid.cell_size[0] * f64::from(col)) + grid.bbox.min().x.to_f64().unwrap();
        let start_height =
            grid.bbox.max().y.to_f64().unwrap() - (grid.cell_size[1] * f64::from(row));
        assert_eq!(
            grid.min_max_points(0, 0, start_width, start_height, grid.inner_cell_size),
            (Point::<f64>::new(0., 0.5,), Point::<f64>::new(0.5, 1.))
        );
        assert_eq!(
            grid.min_max_points(4, 1, start_width, start_height, grid.inner_cell_size),
            (Point::<f64>::new(0.5, -1.5), Point::<f64>::new(1.0, -1.))
        );
    }
}


src/wkt_mapper.rs
use crate::GridGeom;
use geo_types::{Coordinate, Geometry, LineString, MultiPoint, MultiPolygon, Point, Polygon};
use wkt::{
    types::{LineString as WktLineString, Point as WktPoint, Polygon as WktPolygon},
    Geometry as WktGeometry, Wkt,
};

fn map_point(point: WktPoint<f64>) -> Point<f64> {
    let coordinate = point.0.unwrap();
    Point::new(coordinate.x, coordinate.y)
}

fn map_line(linestring: WktLineString<f64>) -> LineString<f64> {
    let coordinates = linestring
        .0
        .into_iter()
        .map(|c| Coordinate::from((c.x, c.y)))
        .collect();
    LineString(coordinates)
}

fn map_polygon(mut polygon: WktPolygon<f64>) -> Polygon<f64> {
    if polygon.0.len() != 1 {
        unimplemented!("Unimplemented: send a PR!")
    }
    let exterior = polygon.0.remove(0);
    let exterior = exterior
        .0
        .into_iter()
        .map(|c| geo_types::Coordinate::from((c.x, c.y)))
        .collect();
    let exterior = LineString(exterior);

    Polygon::new(exterior, vec![])
}

pub fn map(wkt: Wkt<f64>) -> Vec<GridGeom<f64>> {
    wkt.items
        .into_iter()
        .flat_map(|s| match s {
            WktGeometry::Point(point) => {
                GridGeom::<f64>::vec_from_geom(Geometry::Point(map_point(point)), 0.0, false)
            }
            WktGeometry::MultiPoint(points) => {
                let points = points.0.into_iter().map(map_point).collect();
                GridGeom::<f64>::vec_from_geom(Geometry::MultiPoint(MultiPoint(points)), 0.0, false)
            }
            WktGeometry::LineString(linestring) => GridGeom::<f64>::vec_from_geom(
                Geometry::LineString(map_line(linestring)),
                0.0,
                false,
            ),
            WktGeometry::MultiLineString(multistrings) => {
                let multistrings = multistrings.0.into_iter().map(map_line).collect();
                GridGeom::<f64>::vec_from_geom(Geometry::MultiLineString(multistrings), 0.0, false)
            }
            WktGeometry::Polygon(polygon) => {
                GridGeom::<f64>::vec_from_geom(Geometry::Polygon(map_polygon(polygon)), 0.0, true)
            }
            WktGeometry::MultiPolygon(polygons) => {
                let polygons = polygons
                    .0
                    .into_iter()
                    .map(map_polygon)
                    .collect::<Vec<Polygon<f64>>>();
                GridGeom::<f64>::vec_from_geom(
                    Geometry::MultiPolygon(MultiPolygon::from(polygons)),
                    0.0,
                    true,
                )
            }
            _ => unimplemented!("Unimplemented: send a PR!"),
        })
        .collect()
}


src/main.rs
use std::convert::{TryFrom, TryInto};
use std::fs;
use std::io::{self, BufReader, Read};
use std::str::FromStr;
use std::time::Duration;

use anyhow::{self, Context, Result};
use clap::{Arg, ArgAction, Command};
use console::Term;
use geo::{Geometry, Point};
use geojson::{self, FeatureIterator, GeoJson};
use indicatif::ProgressBar;
use kml::{quick_collection, Kml};
use polyline::decode_polyline;
use rstar::RTree;
use topojson::{to_geojson, TopoJson};
use wkt::Wkt;

mod map_grid;
use map_grid::{GridGeom, MapGrid};

#[derive(Debug, PartialEq)]
enum InputFormat {
    GeoJson,
    TopoJson,
    Csv,
    Shapefile,
    Wkt,
    Polyline,
    Kml,
}

impl FromStr for InputFormat {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<InputFormat> {
        match s.to_ascii_lowercase().as_ref() {
            "geojson" => Ok(InputFormat::GeoJson),
            "topojson" => Ok(InputFormat::TopoJson),
            "csv" => Ok(InputFormat::Csv),
            "shp" => Ok(InputFormat::Shapefile),
            "wkt" => Ok(InputFormat::Wkt),
            "polyline" => Ok(InputFormat::Polyline),
            "kml" => Ok(InputFormat::Kml),
            f => Err(anyhow::anyhow!("Invalid format supplied: {}", f)),
        }
    }
}

/// Get file format from flag or infer from file path
fn get_file_format(file_path: &str, file_format: Option<String>) -> Result<InputFormat> {
    let format_str = match file_format {
        Some(f) => f,
        None => file_path.split('.').last().unwrap().to_string(),
    };
    format_str.parse()
}

/// Parse simplification value from float or percentage string
fn get_simplification(simplify: &str) -> Result<f64> {
    if simplify.contains('%') {
        let simplify = simplify
            .replace('%', "")
            .parse::<f64>()
            .with_context(|| format!("Could not parse simplify value: {}", simplify))?;
        Ok(simplify / 100.)
    } else {
        simplify
            .parse::<f64>()
            .with_context(|| format!("Could not parse simplify value: {}", simplify))
    }
}

/// Read file path (or stdin) to string
fn read_input_to_string(file_path: &str) -> Result<String> {
    let mut input_str = String::new();
    match file_path {
        "-" => io::stdin()
            .read_to_string(&mut input_str)
            .context("There was an error reading from stdin"),
        _ => fs::File::open(file_path)
            .unwrap()
            .read_to_string(&mut input_str)
            .with_context(|| format!("There was an error reading from file: {}", file_path)),
    }?;
    Ok(input_str)
}

/// Process top-level GeoJSON items
pub fn process_geojson(gj: GeoJson, simplification: f64, is_area: bool) -> Vec<GridGeom<f64>> {
    match gj {
        GeoJson::FeatureCollection(collection) => collection
            .features
            .into_iter()
            .filter_map(|feature| feature.geometry)
            .flat_map(|g| {
                let geom: Geometry<f64> = g.value.try_into().unwrap();
                GridGeom::<f64>::vec_from_geom(geom, simplification, is_area)
            })
            .collect(),
        GeoJson::Feature(feature) => {
            if let Some(geometry) = feature.geometry {
                let geom: Geometry<f64> = geometry.value.try_into().unwrap();
                GridGeom::<f64>::vec_from_geom(geom, simplification, is_area)
            } else {
                vec![]
            }
        }
        GeoJson::Geometry(geometry) => {
            let geom: Geometry<f64> = geometry.value.try_into().unwrap();
            GridGeom::<f64>::vec_from_geom(geom, simplification, is_area)
        }
    }
}

fn handle_geojson(
    file_path: &str,
    simplification: f64,
    is_area: bool,
) -> Result<Vec<GridGeom<f64>>> {
    Ok(match file_path {
        "-" => FeatureIterator::new(BufReader::new(io::stdin()))
            .filter_map(|f| f.ok())
            .flat_map(|f| process_geojson(GeoJson::Feature(f), simplification, is_area))
            .collect(),
        _ => FeatureIterator::new(BufReader::new(fs::File::open(file_path)?))
            .filter_map(|f| f.ok())
            .flat_map(|f| process_geojson(GeoJson::Feature(f), simplification, is_area))
            .collect(),
    })
}

fn handle_topojson(
    input_str: String,
    simplification: f64,
    is_area: bool,
) -> Result<Vec<GridGeom<f64>>> {
    let topo = input_str
        .parse::<TopoJson>()
        .context("Unable to parse TopoJSON")?;
    match topo {
        TopoJson::Topology(t) => Ok(t
            .list_names()
            .into_iter()
            .map(|n| to_geojson(&t, &n))
            .filter_map(|g| g.ok())
            .map(GeoJson::FeatureCollection)
            .flat_map(|g| process_geojson(g, simplification, is_area))
            .collect()),
        _ => unimplemented!(),
    }
}

fn handle_csv(input_str: String, lat_col: &str, lon_col: &str) -> Result<Vec<GridGeom<f64>>> {
    let mut rdr = csv::Reader::from_reader(input_str.as_bytes());
    let headers = rdr.headers().context("Unable to load CSV headers")?;

    let lat_idx = headers
        .iter()
        .position(|v| v == lat_col)
        .with_context(|| format!("Lat column {} not found", lat_col))?;
    let lon_idx = headers
        .iter()
        .position(|v| v == lon_col)
        .with_context(|| format!("Lon column {} not found", lat_col))?;

    rdr.records()
        .map(|rec_val| {
            let rec = rec_val.context("Could not parse CSV record")?;
            let lat_val: f64 = rec
                .get(lat_idx)
                .unwrap()
                .parse()
                .context("Could not parse lat value from record")?;
            let lon_val: f64 = rec
                .get(lon_idx)
                .unwrap()
                .parse()
                .context("Could not parse lon value from record")?;
            let pt: Point<f64> = Point::new(lon_val, lat_val);
            Ok(GridGeom::Point(pt))
        })
        .collect()
}

fn handle_shp(file_path: &str, simplification: f64, is_area: bool) -> Result<Vec<GridGeom<f64>>> {
    let mut rdr = shapefile::ShapeReader::from_path(file_path)
        .with_context(|| format!("There was an error opening shapefile {}", file_path))?;
    Ok(rdr
        .iter_shapes()
        .filter_map(|s| s.ok())
        .flat_map(|s| match Geometry::<f64>::try_from(s) {
            Ok(geom) => GridGeom::<f64>::vec_from_geom(geom, simplification, is_area),
            Err(_) => vec![],
        })
        .collect())
}

fn handle_wkt(input_str: String, simplification: f64, is_area: bool) -> Result<Vec<GridGeom<f64>>> {
    let wkt = Wkt::<f64>::from_str(&input_str)
        .map_err(|_| anyhow::anyhow!("There was an error parsing WKT"))?;
    let geom: Geometry<f64> = wkt
        .try_into()
        .map_err(|_| anyhow::anyhow!("There was an error converting WKT"))?;
    Ok(GridGeom::<f64>::vec_from_geom(
        geom,
        simplification,
        is_area,
    ))
}

fn handle_polyline(
    input_str: String,
    precision: &str,
    simplification: f64,
) -> Result<Vec<GridGeom<f64>>> {
    let precision: u32 = precision
        .parse()
        .context("Precision has to be defined for polyline format")?;
    let lines = decode_polyline(&input_str, precision).unwrap();
    Ok(GridGeom::vec_from_geom(
        geo_types::Geometry::LineString(lines),
        simplification,
        false,
    ))
}

fn handle_kml(input_str: String, simplification: f64, is_area: bool) -> Result<Vec<GridGeom<f64>>> {
    let kml: Kml = input_str
        .parse()
        .map_err(|_| anyhow::anyhow!("There was an error parsing KML"))?;
    Ok(GridGeom::<f64>::vec_from_geom(
        geo_types::Geometry::GeometryCollection(quick_collection(kml)?),
        simplification,
        is_area,
    ))
}

fn main() -> Result<()> {
    let matches = Command::new(env!("CARGO_PKG_NAME"))
        .version(env!("CARGO_PKG_VERSION"))
        .about(env!("CARGO_PKG_DESCRIPTION"))
        .arg(Arg::new("INPUT")
            .help("File to parse or '-' to read stdin")
            .required(true)
            .index(1))
        .arg(Arg::new("format")
            .short('f')
            .long("format")
            .value_name("FORMAT")
            .help("Input file format (tries to infer from file extension by default)")
            .value_parser(["geojson", "topojson", "csv", "shp", "wkt", "polyline", "kml"])
            .default_value_if("INPUT", "-", Some("geojson")))
        .arg(Arg::new("lon")
            .long("lon")
            .value_name("LON")
            .help("Name of longitude column (if format is 'csv')")
            .default_value("lon"))
        .arg(Arg::new("lat")
            .long("lat")
            .value_name("LAT")
            .help("Name of latitude column (if format is 'csv')")
            .default_value("lat"))
        .arg(Arg::new("rows")
            .short('r')
            .long("rows")
            .value_name("ROWS")
            .help("Sets the number of rows (in characters) of the printed output. Defaults to terminal width."))
        .arg(Arg::new("columns")
            .short('c')
            .long("columns")
            .value_name("COLUMNS")
            .help("Sets the number of columns (in characters) of the printed output. Defaults to terminal height minus 1."))
        .arg(Arg::new("simplify")
            .short('s')
            .long("simplify")
            .help("Proportion of removable points to remove (0-1 or 0%-100%)")
            .default_value("0.01"))
        .arg(Arg::new("precision")
            .long("precision")
            .help("Precision value for polyline parsing")
            .required_if_eq("format", "polyline"))
        .arg(Arg::new("area")
            .short('a')
            .long("area")
            .action(ArgAction::SetTrue)
            .help("Print polygon area instead of boundaries"))
        .get_matches();

    let (term_height, term_width) = Term::stdout().size();
    let height: f64 = match matches.get_one::<String>("rows") {
        Some(ref rows) => rows
            .parse()
            .with_context(|| format!("Rows value {} cannot be parsed as a number", rows)),
        None => Ok(f64::from(term_height - 1)),
    }?;
    let width: f64 = match matches.get_one::<String>("columns") {
        Some(ref cols) => cols
            .parse()
            .with_context(|| format!("Columns value {} cannot be parsed as a number", cols)),
        None => Ok(f64::from(term_width)),
    }?;

    // Simplification is scaled by the output size
    let simplify = get_simplification(matches.get_one::<String>("simplify").unwrap())?;
    let simplification = simplify / (height * width);

    let spinner = ProgressBar::new_spinner();
    spinner.set_message("Reading file");
    spinner.enable_steady_tick(Duration::from_millis(100));
    spinner.set_message("Parsing geography");

    let file_format = get_file_format(
        matches.get_one::<String>("INPUT").unwrap(),
        matches.get_one::<String>("format").cloned(),
    )?;

    let geoms: Vec<GridGeom<f64>> = match file_format {
        InputFormat::GeoJson => handle_geojson(
            matches.get_one::<String>("INPUT").unwrap(),
            simplification,
            matches.get_flag("area"),
        ),
        InputFormat::TopoJson => handle_topojson(
            read_input_to_string(matches.get_one::<String>("INPUT").unwrap())?,
            simplification,
            matches.get_flag("area"),
        ),
        InputFormat::Csv => handle_csv(
            read_input_to_string(matches.get_one::<String>("INPUT").unwrap())?,
            matches.get_one::<String>("lat").unwrap(),
            matches.get_one::<String>("lon").unwrap(),
        ),
        InputFormat::Shapefile => handle_shp(
            matches.get_one::<String>("INPUT").unwrap(),
            simplification,
            matches.get_flag("area"),
        ),
        InputFormat::Wkt => handle_wkt(
            read_input_to_string(matches.get_one::<String>("INPUT").unwrap())?,
            simplification,
            matches.get_flag("area"),
        ),
        InputFormat::Polyline => handle_polyline(
            read_input_to_string(matches.get_one::<String>("INPUT").unwrap())?,
            matches.get_one::<String>("precision").unwrap(),
            simplification,
        ),
        InputFormat::Kml => handle_kml(
            read_input_to_string(matches.get_one::<String>("INPUT").unwrap())?,
            simplification,
            matches.get_flag("area"),
        ),
    }?;

    // Create a combined LineString for bounds calculation
    spinner.set_message("Indexing geography");
    let rtree: RTree<GridGeom<f64>> = RTree::bulk_load(geoms);
    let grid = MapGrid::new(width, height, rtree);
    spinner.finish_and_clear();
    grid.print()
}

#[cfg(test)]
mod test {
    use super::*;
    use geo_types::{Line, Point};

    #[test]
    fn test_get_file_format() {
        assert!(matches!(
            get_file_format("test.GEOJSON", None),
            Ok(InputFormat::GeoJson)
        ));
        assert!(matches!(
            get_file_format("test.geojson", Some("csv".to_string())),
            Ok(InputFormat::Csv)
        ));
    }

    #[test]
    fn test_handle_geojson() {
        let file_path = "./fixtures/input.geojson";
        let outlines = handle_geojson(file_path, 0., false).unwrap();
        let lines = outlines.iter().filter(|g| matches!(g, GridGeom::Line(_)));
        let areas = handle_geojson(file_path, 0., true).unwrap();
        let poly = areas.iter().filter(|g| matches!(g, GridGeom::Polygon(_)));
        assert_eq!(outlines.len(), 14);
        assert_eq!(lines.count(), 13);
        assert_eq!(areas.len(), 5);
        assert_eq!(poly.count(), 3);
    }

    #[test]
    fn test_handle_topojson() {
        let input_str = include_str!("../fixtures/input.topojson").to_string();
        let outlines = handle_topojson(input_str.clone(), 0., false).unwrap();
        let lines = outlines.iter().filter(|g| matches!(g, GridGeom::Line(_)));
        let areas = handle_topojson(input_str, 0., true).unwrap();
        let poly = areas.iter().filter(|g| matches!(g, GridGeom::Polygon(_)));
        assert_eq!(outlines.len(), 14);
        assert_eq!(lines.count(), 13);
        assert_eq!(areas.len(), 5);
        assert_eq!(poly.count(), 3);
    }

    #[test]
    fn test_handle_csv() {
        let input_str = include_str!("../fixtures/input.csv").to_string();
        assert_eq!(
            handle_csv(input_str, "one", "two").unwrap(),
            vec![
                GridGeom::Point(Point::<f64>::new(-1.0, 1.0)),
                GridGeom::Point(Point::<f64>::new(-2.0, 2.0))
            ]
        );
    }

    #[test]
    fn test_handle_shp() {
        let file_path = "./fixtures/input_point.shp";
        let points = handle_shp(file_path, 0., false).unwrap();
        let points = points.iter().filter(|g| matches!(g, GridGeom::Point(_)));
        let file_path = "./fixtures/input_line.shp";
        let lines = handle_shp(file_path, 0., false).unwrap();
        let lines = lines.iter().filter(|g| matches!(g, GridGeom::Line(_)));
        let file_path = "./fixtures/input_area.shp";
        let areas = handle_shp(file_path, 0., true).unwrap();
        let poly = areas.iter().filter(|g| matches!(g, GridGeom::Polygon(_)));
        assert_eq!(points.count(), 1);
        assert_eq!(lines.count(), 1);
        assert_eq!(poly.count(), 3);
    }

    #[test]
    fn test_handle_wkt() {
        let input_str = include_str!("../fixtures/input.wkt").to_string();
        assert_eq!(
            handle_wkt(input_str, 0., false).unwrap(),
            vec![
                GridGeom::Point(Point::<f64>::new(4.0, 6.0)),
                GridGeom::Line(Line::<f64>::new((4.0, 6.0), (7.0, 10.0))),
            ]
        );
    }

    #[test]
    fn test_handle_polyline() {
        let input_str = include_str!("../fixtures/input.polyline.txt").to_string();
        assert_eq!(
            handle_polyline(input_str, "5", 0.).unwrap(),
            vec![
                GridGeom::Line(Line::new((-120.2, 38.5), (-120.95, 40.7))),
                GridGeom::Line(Line::new((-120.95, 40.7), (-126.453, 43.252)))
            ]
        );
    }

    #[test]
    fn test_handle_kml() {
        let input_str = include_str!("../fixtures/input.kml").to_string();
        assert_eq!(
            handle_kml(input_str, 0., false).unwrap(),
            vec![
                GridGeom::Line(Line::new((-1., 2.), (-1.5, 3.))),
                GridGeom::Line(Line::new((-1.5, 3.), (-1.5, 2.))),
                GridGeom::Line(Line::new((-1.5, 2.), (-1., 2.)))
            ]
        );
    }
}
